packages/stock-location/src/services/stock-location.ts

```diff
@@ -167,7 +160,7 @@ export default class StockLocationService extends TransactionBaseService {
     stockLocationId: string,
     updateData: UpdateStockLocationInput
   ): Promise<StockLocation> {
-    return await this.atomicPhase_(async (manager) => {
+    return await this.manager_.transaction(async (manager) => {
```

Comment:
todo: Here we have a new manager for the transaction. But all the other methods called are not aware of that manager and therefore will run outside of the transaction. (retrieve, updateAddress) should those methods accept a `context` argument that optionally include a manager?

Reply:
Same everywhere in the service

Reply:
Here the module is isolated, and now even shared modules won't share DB transactions with the core or other modules.
So I don't think it would change much here. But maybe it worth extending `TransactionBaseService` and if someone is using the module programatically, they could apply the withTransaction to call the methods. And here we revert it to use `atomicPhase_`. wdyt?

Reply:
But should we start to use that way, or continue to use `.withTransaction`?
I'm honestly not a big fan of sharing a context to every single method we call, but if we stick to the withTransaction method, we could just extend `TransactionBaseService`.

Reply:
Let’s go with transactionBaseService this time around. Will see in the future with the transaction management how we plan to do. But the choosen path was to always have a context argument in which you can have things like manager, loggedInUser and stuff like that and that way it is explicit and does not depend on the class anymore

Reply:
If we go with that, it means that the service must be called with withTransaction() without passing the manager, to get a new instance and avoid the shared manager issue.

Reply:
the core won't call any module with `withTransaction` anymore. So that will be only useful if the modules are used in custom apis or other extensions.

Reply:
@carlos-r-l-rodrigues have you been able to look at the last comment?

Reply:
What is you suggestion then? return the `transaction()` as it was before?

Reply:
Without the transaction management being merged, my suggestion would look like:

```ts
async update(
    stockLocationId: string,
    updateData: UpdateStockLocationInput,
    context: { manager?: EntityManager } = {}
  ): Promise<StockLocation> {
    const update = async (manager) => {
      const context_ = { manager }
      const locationRepo = manager.getRepository(StockLocation)

      const item = await this.retrieve(stockLocationId, {}, context_)

      const { address, ...data } = updateData

      if (address) {
        if (item.address_id) {
          await this.updateAddress(item.address_id, address, context_)
        } else {
          const locAddressRepo = manager.getRepository(StockLocationAddress)
          const locAddress = locAddressRepo.create(address)
          const addressResult = await locAddressRepo.save(locAddress)
          data.address_id = addressResult.id
        }
      }

      const { metadata, ...fields } = data

      const toSave = locationRepo.merge(item, fields)
      if (metadata) {
        toSave.metadata = setMetadata(toSave, metadata)
      }

      await locationRepo.save(toSave)

      await this.eventBusService_
        .withTransaction(manager)
        .emit(StockLocationService.Events.UPDATED, {
          id: stockLocationId,
        })

      return item
    }

    if (context?.manager) {
      return await update(context.manager)
    }

    return await this.manager_.transaction(async transactionManager => {
      return await update(transactionManager)
    })
  }
```

wdyt? In that case, no need to extends TransactionBaseService, we can still work with the singleton and each subsequent call is part of the existing transaction or a new transaction that is scoped to the actual running steps

Reply:
I don't like much this shared `context` to call methods. This is not super intuitive and it doesn't change much from keeping the `withTransaction` currently used everywhere.
I will keep the `TransactionBaseService` for now and let this decision for later, as it will only affect whoever is extending the module's class.

Reply:
The problem with extending the TransactionBaseService is that service is not called with the `withTransaction(manager)` we will have some errors `QueryRunnerAlreadyReleased`. You told me that this service will not be called with `withTransction` right? Plus, the shared context has been validated already but we can re discuss if you want. It also makes it very clear that you can or not pass a manger to be used and it remove the manager being hold by the class

Reply:
As far as I can understand, the issue here is, that removing the `atomicPhase` and using `manager.transaction` directly will lead to a new transaction being initiated in every single service call - inside the StockLocationService. 

We don't have a mechanism on the class to ensure that if the transaction was initiated in a previous method within the same service, it will be reused.

If we don't want to add a context param to be passed around, and we don't call the StockLocationService with `withTransaction` from within the core, we need to add a mechanism similar to the atomic phase + transaction base class, such that we reuse transaction managers inside module services.

Reply:
The updateAdress is an example of another write that is not wrapped in the same transaction as the update when not extending the transaction base service. And when extending the transaction base service, the issue is not that. It is that if the service is used concurrently, the first consumer will set the manager on the class and the second one will be executed in the transaction of the first one. Also, if someone want to update the service and add another write within the transaction, it does not seem to be very explicit that he will have to add the mechanism to support it. Wdyt?

Reply:
Another point, the retrieve is originally outside of the transaction, which means that you can end up trying to merge the data of an item that does not exists within the open transaction, am I right?

Reply:
right, which should be using `.withTransaction` to be called.
`await this.withTrasaction(manager).updateAddress(item.address_id, address)`

I'm ok to move that to a new context parameter. I'm just not confortable making this change in one place, where everywhere else is not like this.

Reply:
That will not work if the initiator does not call the service with withTransaction..

Reply:
Discussion resolved here https://medusacorp.slack.com/archives/C04QGK1Q8C8/p1678712394648899?thread_ts=1678710918.025129&cid=C04QGK1Q8C8